core/jobs/io/gcs_io.py

```diff
@@ -0,0 +1,198 @@
+# coding: utf-8
+#
+# Copyright 2022 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Provides an Apache Beam API for operating on GCS."""
+
+from __future__ import annotations
+
+from core.platform import models
+from core.platform.storage import cloud_storage_emulator
+
+import apache_beam as beam
+from typing import List, Optional, Tuple, TypedDict, Union
+
+MYPY = False
+if MYPY:  # pragma: no cover
+    from mypy_imports import app_identity_services
+    from mypy_imports import storage_services
+
+storage_services = models.Registry.import_storage_services()
+app_identity_services = models.Registry.import_app_identity_services()
+
+
+# TODO(#15613): Here we use MyPy ignore because of the incomplete typing of
+# apache_beam library and absences of stubs in Typeshed, forces MyPy to
+# assume that PTransform class is of type Any. Thus to avoid MyPy's error (Class
+# cannot subclass 'PTransform' (has type 'Any')), we added an ignore here.
+class ReadFile(beam.PTransform): # type: ignore[misc]
+    """Read files form the GCS."""
+
+    def expand(self, file_paths: beam.PCollection) -> beam.PCollection:
+        """Returns PCollection with file data.
+
+        Args:
+            file_paths: PCollection. The collection of filepaths that will
+                be read.
+
+        Returns:
+            PCollection. The PCollection of the file data.
+        """
+        return (
+            file_paths
+            | 'Read the file' >> beam.Map(self._read_file)
+        )
+
+    def _read_file(self, file_path: str) -> Tuple[str, Union[bytes, str]]:
+        """Helper function to read the contents of a file.
+
+        Args:
+            file_path: str. The name of the file that will be read.
+
+        Returns:
+            data: Union[bytes, str]. The file data.
+        """
+        bucket = app_identity_services.get_gcs_resource_bucket_name()
+        data = storage_services.get(bucket, file_path)
+        return (file_path, data)
+
+
+class FileObjectDict(TypedDict):
+    """Dictionary representing file object that will be written to GCS."""
+
+    filepath: str
+    data: Union[bytes, str]
+
```

Comment:
Can we now make sure that we always púass bytes to write?

Reply:
I have updated this part and I think now mypy will be able to handle the incoming values.